src/fold.rs

```diff
@@ -0,0 +1,984 @@
+// Adapted from libsyntax.
+
+//! A Folder represents an AST->AST fold; it accepts an AST piece,
+//! and returns a piece of the same type. So, for instance, macro
+//! expansion is a Folder that walks over an AST and produces another
+//! AST.
+//!
+//! **Note**: using a `Folder` on an AST before macro expansion might be
+//! problematic. For example, a folder renaming item names in a module will miss
+//! all of those that are created by the expansion of a macro. Instead, _first_
+//! apply the [`MacExpander`]() `Folder` to expand all the macros (and obtain an
+//! AST without macros), and then apply your `Folder` on that.
+//!
+//! **Warning**: To catch this _very common_ mistake, the default `Folder`
+//! implementation **panics by default on unexpanded macros!** (see
+//! `Folder::fold_mac`). To fold an AST with unexpanded macros, like when
+//! implementing the [`MacExpander`]() folder, override the `Folder::fold_mac`
+//! implementation, for example, using `noop_fold_mac`.
+
+use super::*;
+
+/// AST->AST fold.
+///
+/// Each method of the Folder trait is a hook to be potentially overridden. Each
+/// method's default implementation recursively visits the substructure of the
+/// input via the `noop_fold` methods, which perform an "identity fold", that
+/// is, they return the same structure that they are given (for example the
+/// `fold_crate` method by default calls `fold::noop_fold_crate`).
+///
+/// If you want to ensure that your code handles every variant explicitly, you
+/// need to override each method and monitor future changes to `Folder` in case
+/// a new method with a new default implementation gets introduced.
+pub trait Folder: Sized {
```

Comment:
I think this is just a matter of accepting `<F: ?Sized + Folder>` in the noop functions. I will see if I can make it work.